Add a factory class for customizing the services used by the kubernetes backend plugin

livetocode opened the pull request:
## Hey, I just made a Pull Request!

I refactored the code in order to introduce a factory class for making it easy to replace any service used by the backend plugin.
Previously, we could only provide a KubernetesClustersSupplier but no KubernetesServiceLocator!

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

livetocode commented:
sorry, but I don't understand why there is a merge conflict since I just did a rebase!

Rugvip commented on plugins/kubernetes-backend/api-report.md:
```diff
@@ -151,6 +209,8 @@ export interface RouterOptions {
   // (undocumented)
   config: Config;
   // (undocumented)
+  factory?: KubernetesFactory;
```
Generally we'd pass in a concrete implementation here rather than a factory, do you think that would make sense? Since there is a bit of wiring to do we could get inspired by the [catalog](https://github.com/backstage/backstage/blob/d39cadd9a1dbd8fa2c51deafcf2430df7fb42e3e/packages/backend/src/plugins/catalog.ts#L25), which now has a builder that returns the actual router.

I'm thinking we should also break things down a bit so that you for example only supply a single `KubernetesServiceLocator` implementation when one is provided through code, but if we're constructing things from config we can support the set of hardcoded ones.

livetocode replied:
I refactored the code to introduce a builder and moved the factory methods inside the builder, still allowing one to provide the services through the router options.
Also, I kept the createPlugin function to avoid any breaking change.
Tell me if you like it better!!!!

Rugvip reviewed (commented):
Thank you, I like this direction! 👍

The conflicts seem to be because it's based on an old master. I'd guess you need to sync master in your fork with the main repo, and then rebase

livetocode commented:
Thanks for the quick and positive feedback!
I fixed the merge issues. It was really weird since I didn't notice that the master and origin/master tags in my repo were out of sync! And thus, doing the following commands didn't change anything!
```shell
git checkout master
git pull
git checkout myBranch
git rebase master
```

livetocode commented:
Also, I don't understand why the build keeps failing on the api-docs generation, within plugins that I didn't touch!

See:
```
// Warnings were encountered during analysis:
//
// src/home/components/EntityListDocsTable.d.ts:11:5 - (ae-forgotten-export) The symbol "columnFactories" needs to be exported by the entry point index.d.ts
// src/home/components/EntityListDocsTable.d.ts:12:5 - (ae-forgotten-export) The symbol "actionFactories" needs to be exported by the entry point index.d.ts
// src/plugin.d.ts:29:5 - (ae-forgotten-export) The symbol "TabsConfig" needs to be exported by the entry point index.d.ts
```
And running "yarn build:api-reports:only" on my machine won't help!

The only way to have error messages is to add the --ci argument such as "yarn build:api-reports:only --ci"!
And then I have errors in other plugins than the one listed above!

Please help!

Rugvip commented on plugins/techdocs/api-report.md:
```diff
@@ -118,11 +118,11 @@ export const DocsTable: ({
   actions?:
     | (
         | Action<DocsTableRow>
+        | ((rowData: DocsTableRow) => Action<DocsTableRow>)
```
You'll want to just drop this change from the PR, should be fine

livetocode replied:
Thanks! I didn't notice that I had included this change in my commits!

Rugvip reviewed (commented):
Thanks!

So we'll want the builder interface to look a bit like [freben's proposal](https://github.com/backstage/backstage/pull/6755#issuecomment-898449510), with a number of public `set*` methods for each of the different top-level interfaces, and then a `build()` method to get the result. If the `set*` methods aren't called, the builder should read config and select one of the built-in implementations instead. We'll also want to keep the additions to the public API slim, so only exposing the methods on the builder that we need, and probably keeping the factory private if we use it as an internal helper

livetocode commented:
So, I've adjusted the PR to better mimic the CatalogBuilder example and I hope that you'll like it!

I only kept the build method public and changed the other ones to protected, in case we would subclass the builder.

I also deprecated the createRouter function in favor of the KubernetesBuilder, and I've adjusted the packages/backend kubernetes plugin and the documentation example.

Finally, I didn't change it yet but I have the feeling that we should make services only use a KubernetesClustersSupplier and not a list of hardcoded ClusterDetails[], sunch as implementations of KubernetesServiceLocator, as it would allow them to eventually received an updated list of clusters, especially when GkeClusterLocator is used, since new GKE clusters might be added regularly!
What do you think about it?

livetocode commented:
Ping!

livetocode commented:
I rebased and fixed the merge conflicts.

livetocode commented:
Tks a lot to the team for the quality of review and for your dedication to the project!
